### PR TITLE
Change error to warning in spatialDistribution

### PR DESCRIPTION
**Pull request by AnHeuermann**
### Related Issues
[Track ticket #5281](https://trac.openmodelica.org/OpenModelica/ticket/5281)
https://github.com/OpenModelica/OpenModelica/issues/5281

### Purpose

We are getting errors due to wrong event handling which will start a event search with bisection method which jumps back and forth. `spatialDistribution` was throwing errors when the measured direction didn't match the Boolean `isPositiveVelocity`.

### Approach

If measured direction doesn't match direction given by Boolean just flip it.
If such a value has to be stored in `storeSpatialDistribution` issue a warning, but continue.

### CODE REVIEW

**Comment by casella** [comment]
@AnHeuermann, isn't there a global flag or something that is turned on during bisection? Couldn't we use that to suppress the error message, rather than always making it a warning?

**Comment by AnHeuermann** [comment]
No, I don't think so. We have some variables indicating if we are in event mode, but here we are still searching for the event.

But I can of course remove the annoying warning. Would make more sense then adding a variable somewhere to indicate that we are searching for an event.

**Comment by casella** [comment]
> No, I don't think so. We have some variables indicating if we are in event mode, but here we are still searching for the event.

OK, let me try again. 

I understand this diagnostic is comparing the sign of the velocity with the difference between the last two values of x. This doesn't make sense during bisection search of the zeros of event triggering function, but also, e.g., in case one uses a Runge-Kutta order 5 algorithm, that requires to compute points back and forward in time.

Shouldn't we check the difference between the value of `x` at the end of the previously accepted step with the current one, rather than comparing the last two computed values?

> But I can of course remove the annoying warning.

My point is, we shouldn't warn people if there is nothing to worry about. That is just very confusing.

**Comment by AnHeuermann** [comment]
>I understand this diagnostic is comparing the sign of the velocity with the difference between the last two computed values of x. This doesn't make sense during bisection search of the zeros of event triggering function, but also, e.g., in case one uses a Runge-Kutta order 5 algorithm, that requires to compute points back and forward in time.

Well in the case of bisection we are evaluating `(out0, out1) = spatialDistribution(in0,in1, isPosDirection, posX)` multiple times, but don't store any values. That is only done when a step was accepted by the solver (e.g. DASSL). So nothing to worry about and I removed the error/warning in that case.

>Shouldn't we check the difference between the value of x at the end of the previously accepted step with the current one, rather than comparing the last two computed values?

Yes, and that is what is done in https://github.com/OpenModelica/OpenModelica/pull/7510/files#diff-8527068e943ac83b9544c4be3683f11bde3f7e08524ff3d52207a4edd2b517ddR277-R291
And in that case we get this warning. I guess because the solver did some event iteration but the value of `isPosDirection` is not updated before `storeSpatialDistribution` is called.

So in the end it boils down to the not correctly handled events for spatialDistribution. So maybe I should just remove the warning. There is not much that a user can do about it any way besides not using spatialDistribution in the first place.

**Comment by casella** [comment]
LGTM. @kabdelhak, any chance you can review the code before we merge it in?

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c [hunk 1]
```diff
@@ -272,6 +274,22 @@ void storeSpatialDistribution(DATA* data, threadData_t *threadData, unsigned int
...
+  if (posX - spatialDistribution->oldPosX > 0) {
+    realDirection = 1 /* positive */;
+  } else if (posX - spatialDistribution->oldPosX < 0) {
+    realDirection = -1 /* negative */;
+  } else {
+    realDirection = 0 /* standing still */;
+  }
+
+  /* Check if real dirrection matches isPositiveVelocity */
+  if ((isPositiveVelocity && realDirection==-1 && deltaX > SPATIAL_ZERO_DELTA_X) || (!isPositiveVelocity && realDirection!= -1 && deltaX > SPATIAL_ZERO_DELTA_X)) {
```
Comment (kabdelhak) [inline OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:288]: I don't really understand this check. As far as i understand both `isPositiveVelocity ` and `realDirection` can be -1, 0 or 1.

If you just check if the value itself as condition, it just checks if it is nonzero right? Why do you not check for `isPositiveVelocity == realDirection && deltaX > SPATIAL_ZERO_DELTA_X`? Is it correct that you allow the combinations
(0,1) and (0,-1) to return true? Also shouldn't it also return true if `deltaX < SPATIAL_ZERO_DELTA_X`?

Comment (kabdelhak) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:288]: If any of them can be zero and it is supposed to return true i would simply do `!((isPositiveVelocity * realDirection) < 0) `

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c [hunk 2]
```diff
@@ -272,6 +274,22 @@ void storeSpatialDistribution(DATA* data, threadData_t *threadData, unsigned int
...
+    // TODO: This is probably still a sign that we didn't handle some event or event search correctly.
+    isPositiveVelocity = isPositiveVelocity ? 0:1;
```
Comment (kabdelhak) [inline OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:290]: I also do not understand this manipulation, you map (-1, 0, 1) to (0, 1, 0), but why?

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c [hunk 3]
```diff
@@ -365,10 +384,11 @@ double spatialDistribution(DATA* data, threadData_t *threadData, unsigned int in
     realDirection = 0 /* standing still */;
   }
 
-  /* What should happen if realDirection doesn't match given isPositiveVelocity and velocity is not 0? */
+  /* If real direction doesn't match isPositiveVelocity just flip it.
+   * This still indicates something wrong, so we don't extrapolate the output */
   if ((isPositiveVelocity && realDirection==-1 && deltaX > SPATIAL_ZERO_DELTA_X) || (!isPositiveVelocity && realDirection!= -1 && deltaX > SPATIAL_ZERO_DELTA_X)) {
-    errorStreamPrint(LOG_STDOUT, 0, "Boolean isPositiveDirection doesn't match with direction x is moving.");
-    omc_throw_function(threadData);
+    isPositiveVelocity = isPositiveVelocity ? 0:1;
```
Comment (kabdelhak) [inline OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:390]: Same here... (-1,0,1) to (0,1,0). Why?

**Review by kabdelhak — CHANGES_REQUESTED** [review]
The last part looks good to me (jumped etc.), but i left some comments on stuff i don't really get.

**Comment by casella** [comment]
@AnHeuermann could you please reply to @kabdelhak's comments, so we can go ahead with the commit?

Thanks!

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c (hunk 1 above)
Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:288]: The condition should hold true if `isPositiveVelocity` doesn't match `realDirection`. If `realDirection=0` we don't care

So yes, it would be the same to do
```C
  if (deltaX > SPATIAL_ZERO_DELTA_X && isPositiveVelocity*realDirection > 0 ) {
    isPositiveVelocity = isPositiveVelocity ? 0:1;
  }
```
but is this more easy to understand? Anyway I'll add it since it is shorter.

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c (hunk 2 above)
Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:290]: `isPositiveVelocity` represents a Boolean (so everything that is not `0` is true). We just flip it here.

You could write the same as:
```C
if (isPositiveVelocity != 0)
  isPositiveVelocity = 0
else 
  isPositiveVelocity  = 1
```

Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:290]: Even better would be to do
```C
isPositiveVelocity  = !isPositiveVelocity;
``` I guess

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c (hunk 3 above)
Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:390]: Changed to `isPositiveVelocity  = !isPositiveVelocity;`

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c (hunk 2 above)
Comment (kabdelhak) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c]: ah yes okay i totally was misled by the fact that realDirection is not a boolean. 

Nevermind i am good with this!

**Review by kabdelhak — APPROVED** [review]
LGTM

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c [hunk 4]
```diff
@@ -272,6 +274,23 @@ void storeSpatialDistribution(DATA* data, threadData_t *threadData, unsigned int
...
+  /* Get deltaX */
+  deltaX = spatialDistribution->oldPosX - posX;
+  if (deltaX > 0) {
+    realDirection = 1 /* positive */;
+  } else if (deltaX < 0) {
+    realDirection = -1 /* negative */;
+  } else {
+    realDirection = 0 /* standing still */;
+  }
+  deltaX = fabs(deltaX);
```
Comment (phannebohm) [inline OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:286]: You can even save the `fabs` call while you're at it :smile: 
```suggestion
    realDirection = -1 /* negative */;
    deltaX = -deltaX;
  } else {
    realDirection = 0 /* standing still */;
  }

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c (hunk 2 above)
Comment (casella) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c]: > 
> 
> Even better would be to do
> 
> ```c
> isPositiveVelocity  = !isPositiveVelocity;
> ```

I guess that also applies to `isPositiveVelocity = isPositiveVelocity ? 0:1;`, that I found a bit odd at first sight :)

File: OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c (hunk 4 above)
Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/simulation/solver/spatialDistribution.c:286]: Yes... I could have done that.

**Comment by casella** [comment]
Thanks @AnHeuermann!

I'll double check with tomorrow's nightly and then take care of the trac ticket.
